src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs

```diff
@@ -1803,188 +1803,270 @@ public static bool Decompose(Matrix4x4 matrix, out Vector3 scale, out Quaternion
         {
             bool result = true;
 
-            unsafe
+            float det;
+
+            Matrix4x4 matTemp = Identity;
+            Vector3 canonicalBasis_Row0 = Unsafe.As<float, Vector3>(ref matTemp.M11);
+            Vector3 canonicalBasis_Row1 = Unsafe.As<float, Vector3>(ref matTemp.M21);
+            Vector3 canonicalBasis_Row2 = Unsafe.As<float, Vector3>(ref matTemp.M31);
+            Vector3 pVectorBasis_0 = Unsafe.As<float, Vector3>(ref matrix.M11);
+            Vector3 pVectorBasis_1 = Unsafe.As<float, Vector3>(ref matrix.M21);
+            Vector3 pVectorBasis_2 = Unsafe.As<float, Vector3>(ref matrix.M31);
+
+            translation = Unsafe.As<float, Vector3>(ref matrix.M41);
+
+            scale = new Vector3(pVectorBasis_0.Length(), pVectorBasis_1.Length(), pVectorBasis_2.Length());
+            float pfScales_0 = scale.X, pfScales_1 = scale.Y, pfScales_2 = scale.Z;
+
+            uint a, b;
+            float x = pfScales_0, y = pfScales_1, z = pfScales_2;
+            float pfScales_A, pfScales_B, pfScales_C;
+            Vector3 pVectorBasis_A, pVectorBasis_B, pVectorBasis_C;
+            Vector3 canonicalBasis_A;
+            #region Ranking
```

Comment:
We explicitly don't want to use regions in method bodies. A comment annotating the range here would be better.